archunit/src/main/java/com/tngtech/archunit/library/modules/ArchModules.java

```diff
@@ -0,0 +1,556 @@
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import com.tngtech.archunit.PublicAPI;
+import com.tngtech.archunit.base.ForwardingCollection;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.PackageMatcher;
+import com.tngtech.archunit.library.dependencies.Slices;
+import com.tngtech.archunit.library.modules.ArchModule.Identifier;
+import com.tngtech.archunit.library.modules.ArchModules.Creator.WithGenericDescriptor;
+
+import static com.google.common.base.Functions.toStringFunction;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Multimaps.asMap;
+import static com.tngtech.archunit.PublicAPI.Usage.ACCESS;
+import static com.tngtech.archunit.PublicAPI.Usage.INHERITANCE;
+import static com.tngtech.archunit.core.domain.PackageMatcher.TO_GROUPS;
+import static com.tngtech.archunit.library.modules.ArchModule.Identifier.ignore;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * A collection of {@link ArchModule "architectural modules"}. This class provides a convenient API to partition the {@link JavaClass classes}
+ * of a code base into (cohesive) modules and assert properties of these modules. E.g. their dependencies to each other or dependencies
+ * not contained in any of the constructed modules.<br>
+ * This class provides several entry points to create {@link ArchModule modules} from a set of {@link JavaClass classes} in a convenient way:<br>
+ * <ul>
+ *     <li>{@link #defineBy(IdentifierAssociation)} - the most generic API allowing a lot of flexibility</li>
+ *     <li>{@link #defineByPackages(String)} - an API similar to {@link Slices#matching(String)}</li>
+ * </ul>
+ */
+@PublicAPI(usage = ACCESS)
+public final class ArchModules<DESCRIPTOR extends ArchModule.Descriptor> extends ForwardingCollection<ArchModule<DESCRIPTOR>> {
+    private final Map<Identifier, ArchModule<DESCRIPTOR>> modulesByIdentifier;
+    private final Map<String, ArchModule<DESCRIPTOR>> modulesByName;
+
+    private ArchModules(Set<ArchModule<DESCRIPTOR>> modules) {
+        this.modulesByIdentifier = modules.stream().collect(toMap(ArchModule::getIdentifier, identity()));
+        this.modulesByName = modules.stream().collect(toMap(ArchModule::getName, identity()));
```

Comment:
Yeah, I had thought maybe that together with the value would be good enough, but probably you're right and we should present them in a better way to the user :thinking:

Reply:
Actually, for identifier it doesn't matter. Because the code around ensures there is no way (we're always creating this by grouping classes by identifier, thus 2 modules with the same identifier can't be created).
For name I'm gonna add some validation though...